Add `/ingester/unregister-on-shutdown` HTTP endpoint (#7739)
grafana/mimir: docs/sources/mimir/references/http-api/index.md

This API endpoint is usually used by scale down automations.

### Prepare for unregister

```

> GET,PUT,DELETE /ingester/ingester/unregister-on-shutdown

Author: Good catch, fixed!

```

> This endpoint is used to control whether an ingester should unregister from the ring on its next termination (i.e., the next time it receives a `SIGINT` or `SIGTERM` signal).

Reviewer: You've used i.e. correctly here but it's often confused so Grafana's style guide (an extension of Google's) encourages an alternative.

https://developers.google.com/style/word-list#ie

I took the chance to avoid some passive voice (is used).

```suggestion
This endpoint controls whether an ingester should unregister from the ring on its next termination, that is, the next time it receives a `SIGINT` or `SIGTERM` signal.
```

Via this endpoint, Mimir operators can dynamically control an ingester's `-ingester.ring.unregister-on-shutdown` state without having to restart the ingester.

A `PUT` sets the ingester's unregister state. When invoked with the `PUT` method, the endpoint takes a request body:

```
{"unregister": true}
```

A `GET` returns the ingester's current unregister state.

A `DELETE` resets the ingester's unregister state to the value that was passed via the `-ingester.ring.unregister-on-shutdown`
configuration option.

Regardless of the HTTP method used, the endpoint always returns a response body with the ingester's current unregister state:

```
{"unregister": true}

> ```

Author: > Does this endpoint always return 200?

The endpoint is not guaranteed to always return `200`. It can also return at least these HTTP statuses:
* `503 Service Unavailable` - If the endpoint is called when the ingester is not running.
* `400 Bad Request` - If the JSON sent to the endpoint is malformed.
* `500 Internal Server Error` - If there was an unspecific error decoding the JSON sent to the endpoint.

I've deliberately not included the different HTTP codes in the endpoint's documentation in an attempt to be consistent with the documentation of other endpoints (that don't include detailed information about HTTP response codes).

Thoughts?

> ```

Reviewer 2: Makes sense, thank you. Let's keep it as-is.
